### PR TITLE
[JBPM-9717] Timer instance incorrectly removed upon process execution error

### PR DESCRIPTION
**Pull request by elguardian**
during rollback transaction the timer were incorrecly removed

Jira https://issues.redhat.com/browse/JBPM-9717

### CODE REVIEW

File: jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java [hunk 1]
```diff
@@ -215,13 +218,27 @@ public void clearProcessInstances() {
...
             TimerManager timerManager = ((InternalProcessRuntime)kruntime.getProcessRuntime()).getTimerManager();
-            
+            TimerService timerService = timerManager.getTimerService();
+            if(timerService instanceof RegisteredTimerServiceDelegate) {
+                timerService = ((RegisteredTimerServiceDelegate) timerService).getTimerService();
+            }
+            if(timerService instanceof GlobalTimerService) {
+                // if the timer instance is transactional we don't need to clean up the timers as the rollback part of the transaction awareness
+                // is happening
+                if(!((GlobalTimerService) timerService).isTransactional()) {
```
Comment (martinweiler) [inline jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java:235]: @elguardian shouldn't this method return if the timerService *is* actually transactional? If yes, the ! negation should be removed here imo.

Comment (elguardian) [reply jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java:235]: well... in spite it seems. the answer is no. When is not transactinal the clean up is done in another place (interceptor).
This is a bit strange. This clean up was done 9 years ago while more modern parts (regarding interceptors were done few years later. Also ejb has a flag that is totally out of place (true/false if they can be executed in tx when ejb are always executed within tx. 
Tbh honest this clean up should be removed but at this point it is not really safe.
so this is the best I can do. move the clean up when it is not transactional to the interceptor and remove that situation from here.
This saves the rest of the cases like quartz in spring boot and jdk thread pools

Comment (martinweiler) [reply jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java:235]: I tested this again and it does not work for me. Running in kie-server, isTransactional is true, so the cleanup code is still called.

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

**Review by gmunozfe — APPROVED** [review]
Looks good to me, fabulous work @elguardian !
Just a couple of very minor format fixes, in case you want to update and run CI again...

Tested with different properties (`org.jbpm.ejb.timer.local.cache=false/true` `org.jbpm.ejb.timer.tx=false/true`) and bpmn and works like a charm.

File: jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/tx/NoTransactionalTimerResourcesCleanupAwareSchedulerServiceInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/tx/NoTransactionalTimerResourcesCleanupAwareSchedulerServiceInterceptor.java]: done

File: jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/tx/TransactionAwareSchedulerServiceInterceptor.java [hunk 3]
```diff
@@ -94,8 +98,21 @@ public void beforeCompletion() {
...
+            } else {
+                if (status == TransactionManager.STATUS_ROLLEDBACK) {
+                    // we remove from the timer from the map in timer manager
+                    JobContext ctxorig = timerJobInstance.getJobContext();
+                    ctxorig = (ctxorig instanceof SelfRemovalJobContext) ? ((SelfRemovalJobContext) ctxorig).getJobContext() : ctxorig; // unwrap
+                    if(!(ctxorig instanceof ProcessJobContext)) {
+                        return;
+                    }
+                    Optional<InternalKnowledgeRuntime> runtime = ctxorig.getInternalKnowledgeRuntime();
+                    if(runtime.isPresent()) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/tx/TransactionAwareSchedulerServiceInterceptor.java]: done

**Pull request merged**